File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    
    // 각 작업을 큐에 집어넣는다.
    int length = progresses.size();
    queue<int> q;
    for(int i=0; i<length; i++) q.push(i);
    
    vector<int> answer;
    // 진행도를 차례대로 더한다.
    int count;
    while (!q.empty()) {
        count = 0;
        
        for(int i=0; i<length; i++)
            progresses[i] += (progresses[i] >= 100) ? 0 : speeds[i];
        
        while (!q.empty() && progresses[q.front()] >= 100) {
            q.pop();
            count++;
        }
        
        if(count != 0)
            answer.push_back(count);
    }
    // 큐의 헤드에 위치한 작업이 완료된 경우, poll하는 작업을 반복한다.
    
    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        // 1. 각 작업의 완료일을 계산한다.
        int length = progresses.length;
        int[] days = new int[length];
        for(int i=0; i<length; i++) {
            double div = (double)(100-progresses[i]) / speeds[i];
            days[i] = (int)Math.ceil(div);
        }
        
        // 2. 앞에서 pop하면서, pop한 작업일 수보다 적은 것들을 함께 pop한다.
        List<Integer> ans = new ArrayList<>();
        int count = 1;
        int max_days = days[0];
        for(int i=1; i<length; i++) {
            if (days[i] <= max_days) count++;
            else {
                ans.add(count);
                max_days = days[i];
                count = 1;
            }
        }
        ans.add(count); // 마지막 배포 추가
        
        length = ans.size();
        int[] answer = new int[length];
        int seq = 0;
        for(int i: ans)
            answer[seq++] = i;
        
        return answer;
    }
}